src/backend/parser/parse_relation.c

```diff
@@ -1602,6 +1602,7 @@ addRangeTableEntryForRelation(ParseState *pstate,
 	RangeTblEntry *rte = makeNode(RangeTblEntry);
 	RTEPermissionInfo *perminfo;
 	char	   *refname = alias ? alias->aliasname : RelationGetRelationName(rel);
+	lockmode = get_ENR_withoid(currentQueryEnv, rel->rd_id, ENR_TSQL_TEMP) ? AccessShareLock : lockmode;
```

Comment:
this check should only be done in tsql dialect + if temp oid is enabled, right? lets avoid as much unnecessary overhead as we can

Reply:
I can move the dialect check within the function itself, since the function is called in a number of places anyways. There should be minimal overhead from the function call in that case.

Reply:
Just ran some tests, with 100 threads each creating + dropping 1,000 temp tables.

1. temp oid disabled, default locking: 2:33.621066
2. temp oid disabled, reduced locking: 2:33.507345
3. temp oid enabled, reduced locking: 2:31.988261

So overall there seems to be little performance difference.